openvdb/cmd/openvdb_print.cc

```diff
@@ -11,13 +11,116 @@
 #include <openvdb/openvdb.h>
 #include <openvdb/util/logging.h>
 
+#include <openvdb/points/PointDataGrid.h>
+
+using namespace openvdb;
+using namespace openvdb::points;
 
 namespace {
 
 using StringVec = std::vector<std::string>;
 
 const char* INDENT = "   ";
 const char* gProgName = "";
+static const std::string sINDENT(INDENT);
+
+#if OPENVDB_ABI_VERSION_NUMBER >= 6
+struct PointStats{
+// Helper structs to collect point information
+struct PointAttrib{
+    Name name = "";
+    Name type = "";
+    Name codec = "";
+    Index64 index = 0;
+    bool isUniform = true;
+    bool isShared = false;
+    // flags
+    bool isHidden = false;
+    bool isTransient = false;
+    bool isStreaming = false;
+};
+struct Points{
+    Index64 total = 0;
+    Index64 active = 0;
+    Index64 inactive = 0;
+    std::map<Name, Index64> groups;
+    std::map<Name, PointAttrib> pointAttribs;
+};
+
+static void
+inspectPoints(const openvdb::GridBase::ConstPtr grid, Points& points)
+{
+    PointDataGrid::ConstPtr inputGrid = GridBase::grid<PointDataGrid>(grid);
+
+    for (auto leafIter = inputGrid->tree().cbeginLeaf(); leafIter; ++leafIter) {
+        auto attrset = leafIter->attributeSet();
+        auto dptr = attrset.descriptorPtr();
+        auto attrmap = dptr->map();
+
+        points.total += leafIter->pointCount();
+        points.active += leafIter->onPointCount();
+        points.inactive += leafIter->offPointCount();
+
+        // groups
+        auto grmap = dptr->groupMap();
+
+        // Count points in groups
+        for (auto it=grmap.begin(); it!=grmap.end(); ++it){
+            if(points.groups.find(it->first) == points.groups.end()){
+                points.groups[it->first] = 0;
+            }
+            points.groups[it->first]+=leafIter->groupPointCount(it->first);
+        }
+
+        for (auto it = grmap.begin(); it!=grmap.end(); ++it){
+            std::cout << it->first << std::endl;
+        }
+        // Collect attribute info
+        for (auto it=attrmap.begin(); it!=attrmap.end(); ++it){
+            PointAttrib pa{};
+            pa.name = it->first;
+            pa.index = it->second;
+            pa.type = dptr->valueType(pa.index);
+            pa.isShared = attrset.isShared(pa.index);
+
+            auto attrArr = attrset.getConst(pa.name);
+            pa.codec = attrArr->codecType();
+            pa.isUniform = attrArr->isUniform();
+            pa.isHidden = attrArr->isHidden();
+            pa.isTransient = attrArr->isTransient();
+            pa.isStreaming = attrArr->isStreaming();
+            points.pointAttribs[pa.name] = pa;
+        }
+    }
```

Comment:
I'm not sure how useful this multi leaf implementation is. In most cases the attributes will be consistent across the VDB, and even if they are not, this doesn't provide a way to infer exactly how the attributes are spread. My vote would be to simply inspect the first leaf and infer the attributes from that. You could then check to see fit he descriptor is shared and, if it isn't, simply print a warning that the attribute layouts are not consistent.

Originally comments on this are here https://jira.aswf.io/browse/OVDB-114

@danrbailey